File: 문제풀이/온라인 저지/백준/# 04. Bronze Ⅱ/# 14471. 포인트 카드/java/Main.java
import java.io.DataInputStream;
import java.io.DataOutputStream;

public class Main {

    public static void main(String[] args) {
        int n = i();
        int m = i();
        int success = 0;
        Heap heap = new Heap(m);

        for (int i=0, tmp; i<m; i++) {
            tmp = n- i();
            i();
            if (tmp <= 0) success ++;
            else heap.push(tmp);
        }
        int answer = 0;
        while (success < m-1) {
            answer += heap.pop();
            success ++;
        }
        writeInt(answer);
        flushBuffer();
    }

    private static final byte EOF = -1;
    private static final byte ASCII_minus = 45;
    private static final byte ASCII_0 = 48;
    private static final DataInputStream din = new DataInputStream(System.in);
    private static final DataOutputStream dout = new DataOutputStream(System.out);
    private static final byte[] inbuffer = new byte[65536];
    private static int inbufferpointer;
    private static int bytesread;
    private static final byte[] outbuffer = new byte[65536];
    private static int outbufferpointer;
    private static final byte[] bytebuffer = new byte[20];

    private static int i() {
        int v = 0;
        byte c = r();
        do {v = v * 10 + c - 48;} while (isDigit((c = r())));
        return v;
    }

    private static byte r() {
        if (inbufferpointer == bytesread) fillBuffer();
        return (bytesread == EOF) ? EOF : inbuffer[inbufferpointer++];
    }

    private static void fillBuffer() {
        try {
            bytesread = din.read(inbuffer, (inbufferpointer = 0), inbuffer.length);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static void writeInt(int i) {
        if (i == 0) {
            writeByte(ASCII_0);
        } else {
            if (i < 0) {
                writeByte(ASCII_minus);
                i = -i;
            }
            int index = 0;
            while (i > 0) {
                bytebuffer[index++] = (byte)(i % 10 + ASCII_0);
                i /= 10;
            }
            while (index-- > 0) {
                writeByte(bytebuffer[index]);
            }
        }
    }

    private static void writeByte(byte b) {
        if (outbufferpointer == outbuffer.length) flushBuffer();
        outbuffer[outbufferpointer++] = b;
    }

    private static void flushBuffer() {
        if (outbufferpointer != 0) {
            try {
                dout.write(outbuffer, 0, outbufferpointer);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
            outbufferpointer = 0;
        }
    }
    private static boolean isDigit(byte b) {
        return b >= ASCII_0;
    }
}


class Heap {

    private final int[] arr;
    private int size;

    Heap(int cap) {
        this.arr = new int[cap+1];
    }

    void push(int e) {
        arr[++size] = e;
        int i = size;
        int p = i >> 1;
        int tmp;
        while (p > 0) {
            if (arr[i] < arr[p]) {
                tmp = arr[p];
                arr[p] = arr[i];
                arr[i] = tmp;
                i = p;
                p = i >> 1;
            } else {
                break;
            }
        }
    }

    int pop() {
        int e = arr[1];
        arr[1] = arr[size--];
        int i = 1;
        int lt = i << 1;
        int rt = lt + 1;
        int small = i;
        int tmp;
        while (true) {
            if (lt <= size && arr[lt] < arr[small]) small = lt;
            if (rt <= size && arr[rt] < arr[small]) small = rt;
            if (i == small) break;
            tmp = arr[i];
            arr[i] = arr[small];
            arr[small] = tmp;
            i = small;
            lt = i << 1;
            rt = lt + 1;
        }
        return e;
    }
}

File: 문제풀이/온라인 저지/백준/# 04. Bronze Ⅱ/# 14471. 포인트 카드/kotlin/Main.kt
fun main() {
    val n = i()
    val m = i()
    var success = 0
    val heap = Heap(m)
    repeat(m) {
        (n - i()).also { if (it <= 0) success ++ else heap.push(it) }
        i()
    }
    var answer = 0
    while (success < m-1) {
        answer += heap.pop()
        success ++
    }
    writeInt(answer)
    flushBuffer()
}

class Heap(cap: Int) {

    private val arr = IntArray(cap + 1)
    private var size = 0

    fun push(e: Int) {
        arr[++size] = e
        var i = size
        var p = i shr 1
        while (p > 0) {
            if (arr[i] < arr[p]) {
                arr[p] = arr[i].also { arr[i] = arr[p] }
                i = p
                p = i shr 1
            } else {
                break
            }
        }
    }

    fun pop() : Int {
        val e = arr[1]
        arr[1] = arr[size--]
        var i = 1
        var lt = i shl 1
        var rt = lt + 1
        var small = i
        while (true) {
            if (lt <= size && arr[lt] < arr[small]) small = lt
            if (rt <= size && arr[rt] < arr[small]) small = rt
            if (i == small) break
            arr[i] = arr[small].also { arr[small] = arr[i] }
            i = small
            lt = i shl 1
            rt = lt + 1
        }
        return e
    }
}

private const val EOF = -1
private const val ASCII_minus: Byte = 45
private const val ASCII_0: Byte = 48
private val din = java.io.DataInputStream(System.`in`)
private val dout = java.io.DataOutputStream(System.out)
private val inbuffer = ByteArray(65536)
private var inbufferpointer = 0
private var bytesread = 0
private val outbuffer = ByteArray(65536)
private var outbufferpointer = 0
private val bytebuffer = ByteArray(20)

private fun i(): Int {
    var v = 0
    var c = r()
    do v = v * 10 + c - 48 while (isDigit(r().also { c = it }))
    return v
}

private fun r(): Byte {
    if (inbufferpointer == bytesread) fillBuffer()
    return if (bytesread == EOF) EOF.toByte() else inbuffer[inbufferpointer++]
}

private fun fillBuffer() {
    bytesread = din.read(inbuffer, 0.also { inbufferpointer = it }, inbuffer.size)
}

private fun writeInt(i: Int) {
    var i = i
    if (i == 0) {
        writeByte(ASCII_0)
    } else {
        if (i < 0) {
            writeByte(ASCII_minus)
            i = -i
        }
        var index = 0
        while (i > 0) {
            bytebuffer[index++] = (i % 10 + ASCII_0).toByte()
            i /= 10
        }
        while (index-- > 0) {
            writeByte(bytebuffer[index])
        }
    }
}

private fun writeByte(b: Byte) {
    if (outbufferpointer == outbuffer.size) flushBuffer()
    outbuffer[outbufferpointer++] = b
}

private fun flushBuffer() {
    if (outbufferpointer != 0) {
        dout.write(outbuffer, 0, outbufferpointer)
        outbufferpointer = 0
    }
}

private fun isDigit(b: Byte) = b >= ASCII_0
